bench/bench.ml

```diff
@@ -0,0 +1,506 @@
+  if config.with_metrics then (
+    Metrics.enable_all ();
+    Metrics_gnuplot.set_reporter ();
+    Metrics_unix.monitor_gc 0.1 );
+  bindings_pool := make_bindings_pool config.nb_entries;
+  absent_bindings := make_bindings_pool config.nb_entries
+
+let print fmt (config, results) =
+  let pp_bench fmt (b, result) =
+    Format.fprintf fmt "%s@\n    @[%a@]" b.Index.synopsis Benchmark.pp_result
+      result
+  in
+  Format.fprintf fmt
+    "Configuration:@\n    @[%a@]@\n@\nResults:@\n    @[%a@]@\n" pp_config config
+    Fmt.(list ~sep:(any "@\n@\n") pp_bench)
+    results
+
+let print_json fmt (config, results) =
+  let open Yojson.Safe in
+  let obj =
+    `Assoc
+      [
+        ("config", config_to_yojson config);
+        ( "results",
+          `Assoc
+            (List.map
+               (fun (b, result) ->
+                 (b.Index.name, Benchmark.result_to_yojson result))
+               results) );
+      ]
+  in
+  pretty_print fmt obj
+
+let run filter root output seed with_metrics log_size nb_entries json =
+  let config =
+    { key_size; value_size; nb_entries; log_size; seed; with_metrics }
+  in
+  cleanup root;
+  init config;
+  let name_filter =
+    match filter with None -> fun _ -> true | Some re -> Re.execp re
+  in
+  Index.suite
+  |> schedule name_filter
+  |> List.map (fun (b : Index.suite_elt) ->
+         let name =
+           match b.dependency with None -> b.name | Some name -> name
+         in
+         let result =
+           Index.run ~with_metrics ~nb_entries ~log_size ~root ~name
+             ~fresh:b.fresh ~readonly:b.readonly b.benchmark
+         in
+         (b, result))
+  |> fun results ->
+  let fmt =
+    (match output with None -> stdout | Some filename -> open_out filename)
+    |> Format.formatter_of_out_channel
+  in
+  Fmt.pf fmt "%a@." (if json then print_json else print) (config, results)
+
+open Cmdliner
+
+let env_var s = Arg.env_var ("INDEX_BENCH_" ^ s)
+
+let new_file =
+  let parse s =
+    match Sys.file_exists s with
+    | false -> `Ok s
+    | true -> `Error (Printf.sprintf "`%s' already exists" s)
+  in
+  (parse, Format.pp_print_string)
+
+let regex =
+  let parse s =
+    try Ok Re.(compile @@ Pcre.re s) with
+    | Re.Perl.Parse_error -> Error (`Msg "Perl-compatible regexp parse error")
+    | Re.Perl.Not_supported -> Error (`Msg "unsupported regexp feature")
+  in
+  let print = Re.pp_re in
+  Arg.conv (parse, print)
+
+let name_filter =
+  let doc =
+    "A regular expression matching the names of benchmarks to run. For more \
+     information about the regexp syntax, please visit \
+     https://perldoc.perl.org/perlre.html#Regular-Expressions."
+  in
+  let env = env_var "NAME_FILTER" in
+  Arg.(
+    value
+    & opt (some regex) None
+    & info [ "f"; "filter" ] ~env ~doc ~docv:"NAME_REGEX")
+
+let data_dir =
+  let doc = "Set directory for the data files" in
+  let env = env_var "DATA_DIR" in
+  Arg.(value & opt dir "_bench" & info [ "d"; "data_dir" ] ~env ~doc)
```

Comment:
```suggestion
  Arg.(value & opt dir "_bench" & info [ "d"; "data-dir" ] ~env ~doc)
```

I think it's more conventional to use hyphens for linking words in these options (https://www.gnu.org/prep/standards/html_node/Option-Table.html#Option-Table).